classes/local/rule_instances_controller.php

```diff
@@ -44,90 +44,179 @@
  * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
  */
 class rule_instances_controller implements renderable, \templatable {
-    /** @var stdClass[] rule instance records from the database, sorted by "sortorder" */
+    /** @var stdClass[] external representation of rule instance records, sorted by "sortorder" */
     protected $ruleinstances = [];
 
+    /** @var stdClass[] internal representation of $ruleinstances (used for tracking changes) */
+    protected $ruleinstancesinternal = [];
+
     /**
      * Constructor
      *
      * @throws dml_exception
      */
     public function __construct() {
         global $DB;
-
+        // Get a sorted list of rule instance records from the database.
         $instancerecords = $DB->get_records(
             table: 'tool_registrationrules',
             sort: 'sortorder ASC',
         );
-        $this->ruleinstances = $instancerecords;
+        $this->ruleinstances = $this->ruleinstancesinternal = $instancerecords;
+    }
+
+    /**
+     * Commit all changes to rule instances to the database, sorts the new
+     * list of rule instance records and resets the internal representation.
+     *
+     * @return void
+     */
+    protected function commit(): void {
+        global $DB;
+        // We'll possibly be committing a number of new records and updates
+        // to the databse so best to create a new transaction.
+        $transaction = $DB->start_delegated_transaction();
+        // Iterate over the internal rule insatnces list, determine if we have
+        // anything to commit and perform the appropriate database operation.
+        foreach ($this->ruleinstancesinternal as $internal) {
+            // New instances.
+            if (!isset($internal->id)) {
+                $internal->id = $DB->insert_record('tool_registrationrules', $internal);
+                $this->ruleinstances[$internal->id] = $internal;
+            }
+            // Modified instances.
+            if (isset($internal->modified)) {
+                unset($internal->modified);
+                $DB->update_record('tool_registrationrules', $internal);
```

Comment:
Like for `$DB->insert_record()` we can also set a bulk parameter here.

Reply:
Unfortunately I fear this would do nothing :(